language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/Cek.hs

```diff
@@ -273,34 +292,45 @@ computeCek con t@(Var _ varName)   = do
-returnCek (FrameApplyArg argVarEnv arg : con) fun = do
+returnCek (FrameTyInstArg _ : ctx) fun =
+    case fun of
+      TyAbs _ _ _ body -> computeCek ctx body  -- NB: we're just ignoring the type here
+      _                ->  throwingWithCause _MachineError NonTyAbsInstantiationMachineError $ Just (void fun)
+returnCek (FrameApplyArg argVarEnv arg : ctx) fun = do
     funVarEnv <- getVarEnv
-    withVarEnv argVarEnv $ computeCek (FrameApplyFun funVarEnv fun : con) arg
-returnCek (FrameApplyFun funVarEnv fun : con) arg = do
-    argVarEnv <- getVarEnv
-    applyEvaluate funVarEnv argVarEnv con fun arg
-returnCek (FrameIWrap ann pat arg : con) val =
-    returnCek con $ IWrap ann pat arg val
-returnCek (FrameUnwrap : con) dat = case dat of
-    IWrap _ _ _ term -> returnCek con term
+    withVarEnv argVarEnv $ computeCek (FrameApplyFun funVarEnv fun : ctx) arg
+returnCek (FrameApplyFun funVarEnv fun : ctx) arg =
+    case fun of
+      LamAbs _ name _ body -> do
+               argVarEnv <- getVarEnv
+               withVarEnv (extendVarEnv name arg argVarEnv funVarEnv) $ computeCek ctx body
+      _ ->  throwingWithCause _MachineError NonLambdaApplicationMachineError $ Just (Apply () (void fun) (void arg))
+      -- FIXME: ^ error message looks like eg (1 2) instead of [(con 1) (con 2)]
+returnCek (FrameIWrap ann pat arg : ctx) val =
+    returnCek ctx $ IWrap ann pat arg val
+returnCek (FrameUnwrap : ctx) dat = case dat of
+    IWrap _ _ _ term -> returnCek ctx term
     term             ->
         throwingWithCause _MachineError NonWrapUnwrappedMachineError $ Just (void term)
-
--- | Instantiate a term with a type and proceed.
--- In case of 'TyAbs' just ignore the type. Otherwise check if the term is an
--- iterated application of a 'BuiltinName' to a list of 'Value's and, if succesful,
--- apply the term to the type via 'TyInst'.
-instantiateEvaluate
-    :: (GShow uni, GEq uni, DefaultUni <: uni, Closed uni, uni `Everywhere` ExMemoryUsage)
-    => Context uni -> Type TyName uni ExMemory -> WithMemory Term uni -> CekM uni (Plain Term uni)
-instantiateEvaluate con _ (TyAbs _ _ _ body) = computeCek con body
-instantiateEvaluate con ty fun
-    | isJust $ termAsPrimIterApp fun = returnCek con $ TyInst (memoryUsage () <> memoryUsage fun <> memoryUsage ty) fun ty
-    | otherwise                      =
-        throwingWithCause _MachineError NonPrimitiveInstantiationMachineError $ Just (void fun)
+returnCek (FrameApplyBuiltin initialEnv finalEnv bn tys values terms : ctx) value = do
+    -- See Note [Built-in application] for an explanation of what's going on here.
+    case value of
+      Constant{} -> do
+          let values' = value:values
+          case terms of
+            []       -> applyBuiltin finalEnv ctx bn tys (reverse values') -- We've accumulated the argument closures in reverse
+            arg:args -> withVarEnv initialEnv $ computeCek (FrameApplyBuiltin initialEnv finalEnv bn tys values' args : ctx) arg
```

Comment:
I'd perhaps return `values'` and `newFinalEnv` (why not `finalEnv'` or `newValues` then?) from `case value of` and put a single `case terms of` below `case value of`. Otherwise we duplicate some non-trivial code, which is not very good.

Reply:
Ah yes, I'd meant to go back and look at that. It's a lot neater now.